drafts/inlinemacros.md

```diff
@@ -165,99 +170,66 @@ positions:
-      inline def ++ (y: Int): Int = plus(this.x, y)
-    }
-    inline def plus(x: inline Int, y: inline Int): inline Int = x + y
-    def plus(x: Int, y: Int): Int = x + y
-
-Let's assume we have
-
-    a, b: inline Int
-
-Then we rewrite `a ++ b` as follows:
-
-    a ++ b
-    -->
-    PlusWrapper(a).++(b)
-    -->
-    new PlusWrapper(a).++(b)
-    -->
-    plus(new PlusWrapper(a).x, b)
-    -->
-    plus(a, b)
-    -->
-    a + b: inline Int
-
-Note that the call to `plus` in `PlusWrapper` resolves statically to the non-inline method `plus`.
-Yet in the last-but-one expression of the above rewrite sequence it is rewritten to refer to the
-inline method `plus` because at that point both arguments are known to be inline values.
-
-
 ## Macros
 
-A macro is an expression of the form
+A macro expression is an expression of the form
 
     macro { ... }
 
-where `{ ... }` is some block of Scala code. (In fact, `macro` may prefix arbitrary expressions,
-but blocks are used most commonly). Macros may only appear in the bodies of inline methods.
-
-Macros can only reference the following things in their environment:
-
- 1. parameters of enclosing inline methods (where the parameter may, or may not be, marked inline),
- 2. inline values,
- 3. `this` references to classes containing an enclosing inline method as a member
- 4. anything that's global,
-
-A macro is said to _close over_ items in (1)-(3).
-
-Inside a macro, the type of closed-over definitions changes. If the definition is of type
-`inline T` outside the macro, it is of type `T` inside. If it is a non-inline type `U`, the type in
-the macro is `Expr[U]`, the type of expression trees of type `U`. Types with several levels of
-`inline` are handled analogously. For instance, a type like
-
-    inline (inline Int, Boolean) => String
-
-would be seen inside a macro as
-
-    (Int, Expr[Boolean]) => Expr[String]
-
-A dual transformation is applied to a macro's result. Normal types are prefixed with `inline` until
-a type is of the form `Expr[T]` in which case it is replaced by `T`. So if a macro body returned
-a list of type
-
-    List[(String, Expr[T])]
-
-the type of the macro itself would be
-
-    inline List[inline (inline String, T)]
-
-## Whitebox vs Black Box
-
-Normaly, we would require that an inline method has an explicit result type.
-This restriction is in place so that we can expand inline functions and macros
-in a separate phase after type checking.
-
-We could allow inline functions with missing result type (so called "whitebox macros"),
-maybe controlled by a language import such as
-
-    import language.whiteboxMacros
-
+where `{ ... }` is some block of Scala code, called macro body. (In fact, `macro` may prefix arbitrary expressions,
+but blocks are used most commonly). 
+
+In a macro expression, `macro` stands for a function 
+`def macro[T](body: implicit scala.meta.macros.Context => scala.meta.Term[T]): T = ???` declared in `scala.Predef`.
```

Comment:
The necessity to explicitly specify the type of `Term[T]` will be really annoying, especially since after Scala 2.11 people are accustomed to returning unadorned quasiquotes from their macros, e.g. `def impl(c: Context) = { import c.universe._; q"42" }`. 

Therefore, how about we allow inference of that `T` thing from an expected type at the typecheck site of the macro expression. E.g. `inline def foo: Int = macro { ... }` would infer `T` to be `Int` after the return type of the inline method.